README.md


R Package for Downloading and Handling DWD MOSMIX Forecasts
===========================================================

The german weather service (DWD) publishes a set of automated
weather forecasts (MOS forecasts) as publicly available data
sets.

* [DWD open data server](https://opendata.dwd.de)
* [DWD copyright information](https://www.dwd.de/copyright)
* [DWD MOSMIX product information](https://www.dwd.de/EN/ourservices/met_application_mosmix/met_application_mosmix.html)

This small working package allows to easily download and decode
these forecasts. The forecasts become available several times
a day, XML format. The `mosmix` R package provides the following
features:

* Check files available on the server
* Automatically download and parse forecasts for a set of user-defined
  locations
* Returns convenient time series objects
* Write (subsetted) data sets to local ASCII files

Allows to download _MOSMIX-L_ (4x daily, 115 parameters, up to 240 hours ahead)
and _MOSMIX-S_ (24x daily, 40 parameters, up to 240 hours ahead).

For the decoding of the data sets this package depends on the
[XML](https://cran.r-project.org/package=XML) package.
Plesae note that, to be able to install the R package, `libxml` has to
be installed. The forecasts are time series data wherefore this package also
depends on the R package [zoo](https://cran.r-project.org/package=zoo).


Package Status
==============

Please note that this is a **working package** and was written in less
then a day! Should be seen as an early beta version. Feel free to
use, reuse, improve and contribute!

All best,

_Reto_

Documentation
=============

I started writing a githubio documentation, but currently only a rough
skeleton!

* [retostauffer.github.io/Rmosmix](https://retostauffer.github.io/Rmosmix)

Package Installation
====================

`mosmix` can be installed using the _R_ package `githubinstall` (or `devtools`,
or simply clone and install locally).

## with devtools via https (`devtools::install_github`)

Currently the package is only available via [github](https://github.com).
One convenient way to install packages from [github](https://github.com) is
is via `devtools` (requires _R_ package `devtools` to be installed; if not
use `install.packages("devtools")`).

The following snippet installes the latest _development_ version of the
mosmix package from [github.com/retostauffer/Rmosmix](https://github.com/retostauffer/Rmosmix):

```
# Load devtools
library("devtools")

# Install package, should also resolve the dependencies
install_github("retostauffer/Rmosmix")
```

**Note**: If you do not like experimental versions feel free to install one
of our release candidates, e.g., `v0.0-2`, an early alpha release from
December 2018. To install a release candidate provide the additional `ref`
argument. A list of available
[release candidates](https://github.com/retostauffer/Rmosmix/releases)
can be found on the [github release page](https://github.com/retostauffer/Rmosmix/releases).

```
# Load devtools
library("devtools")

# Install package, should also resolve the dependencies
devtools::install_github("retostauffer/Rmosmix", ref = "v0.0-2")
```

## devtools by cloning the repository

You can also clone the repository and install the package based
on the checkout. For linux users:

```
# Change directory
cd <somewhere/on/your/local/disc>

# Clone repository
git clone https://github.com/retostauffer/Rmosmix.git

# Start R
R
```

As soon as you are in the interactive _R_ shell:

```
# Loading the devtools library
library("devtools")

# Install package
devtools::install("Rmosmix")
```

## Via console

Or of course via console:

```
# Change directory
cd <somewhere/on/your/local/disc>

# Clone repository
git clone https://github.com/retostauffer/Rmosmix.git

# Feel free to use on of our release candidates which
# might miss some features, but might also be more stable.
# A list of release candidates can be found on:
# - https://github.com/retostauffer/Rmosmix/releases
(cd Rmosmix && git checkout tags/<tagname> && cd ..)

# Install directly
R CMD INSTALL Rmosmix

# OR
R CMD build Rmosmix
R CMD INSTALL mosmix_<version>.tar.gz
```




NEWS.md

TODO
----

* Loading all nodes is memory inefficient, try to find a way
  to loop along an xpath.
* Colors for the plotting function.
* Documentation.
* Testing.

Version 0.1
-----------

I've just set up a small R package for downloading handling
and accessing DWD MOSMOX forecasts. Current features:

* Selective download of MOSMIX-L and MOSMIX-S forecasts for
   user-defined set of locations/stations (see details below)
* `get_files_avail` checks the available files on the server.
* `tests/automation.R` contains a small script which might can
   be used for automation. Will, at some point, be put into
   an automation function for handy use. The script
   is checking available files on the server and only processes
   the files which have not yet been downloaded, parsed, and
   stored. This script can be used fur an _archive script_.
   We use it to store some forecasts from the live rolling
   archive for later analysis.

**Some information about the data**: the German weather service
(DWD) provides some DWD MOSMIX forecasts (statistically corrected
forecasts based on the ICON/ECMWF numerical weather prediction model)
on their [opendata server](https://opendata.dwd.de). Two products
are available: 

* MOSMIX-L: produced 4 times a day up to 240 hours ahead, provides
  up to 115 parameters (location dependent).
* MOSMIX-S: produced every hour up to 240 hours ahead. Provides
  up to 40 parameters (location dependent).

MOSMIX-L forecasts are available as individual files, one file
per station. Thus, downloading these forecasts is rather efficient.
For MOSMIX-S one single file is available on the server containing
all locations (several thousand different locations). To be able
to extract one station the whole file has to be downloaded. Thus, the
MOSMIX-M takes much longer to process and needs some more bandwidth,
however, works like a charm!

If you use these data sets please read the DWD terms and conditions
regarding their open data sets! More information:


* [DWD open data server](https://opendata.dwd.de)
* [DWD copyright information](https://www.dwd.de/copyright)
* [DWD MOSMIX product information](https://www.dwd.de/EN/ourservices/met_application_mosmix/met_applicati
on_mosmix.html)




tests/automation.R
# -------------------------------------------------------------------
# - NAME:        automation.R
# - AUTHOR:      Reto Stauffer
# - DATE:        2019-01-09
# -------------------------------------------------------------------
# - DESCRIPTION: Some tests. This script can be used as a template
#                for an automation script. 
# -------------------------------------------------------------------
# - EDITORIAL:   2019-01-09, RS: Created file on thinkreto.
# -------------------------------------------------------------------
# - L@ST MODIFIED: 2019-01-09 19:49 on marvin
# -------------------------------------------------------------------

    library("mosmix")

# -------------------------------------------------------------------
# L-type forecasts: XML files are available for each station
# individually. Thus, we only have to download a few kilobytes of
# data. L-type forecasts become available four times a day.
# -------------------------------------------------------------------

    # List of required stations (user defines station list)
    #  6660: Zuerich
    # 11120: Innsbruck Airport
    # 11035: Vienna, Hohe Warte
    stations <- c(6660, 11120, 11035)

    # Manually create the output folder if not yet existing.
    dir.create("DWDMOS_L", showWarnings = FALSE)

    # Processing L-type forecasts (station-by-station)
    cat("\n\nProcessing L-type DWD MOSMIX forecasts\n")

    # Looping over the user-specified set of stations
    for ( stn in stations ) {

        # Check for available files on the DWD server
        files <- get_files_available("L", stn, outdir = "DWDMOS_L")

        # No files we have not yet processed? Continue.
        if ( nrow(files) == 0 ) next
    
        # Else loop over all available files and process the data.
        for ( i in 1:nrow(files) ) {

            cat(sprintf("* Downloading \"%s\"\n", as.character(files$srcfile[i])))

            # Downloading the data set
            kmz <- tempfile(pattern = "DWDMOS_L_", fileext = ".kmz")
            check   <- try(download.file(as.character(files$url[i]), kmz, quiet = TRUE))
            if ( inherits(check, "try-error") )
                warning(sprintf("Problems downloading data for %d, skip.", stn))

            # Unzip and parse the XML file
            kml      <- unzip(kmz)
            doc      <- xmlParse(kml)

            # Extracting meta information and datetime information
            meta     <- get_meta_info(doc)
            datetime <- get_datetime(doc)

            # Extract forecasts for this station and write to ASCII file.
            fcst <- get_forecasts(sprintf("%05d", stn), doc, datetime, meta)
            write_ascii(fcst, dir = "DWDMOS_L")
            free(doc) # Release connection and free memory

            # Remove the downloaded kmz and extracted kml file.
            file.remove(kmz, kml)
        }
    }


# -------------------------------------------------------------------
# S-type forecasts become available 24 times a day. Only one "global"
# XML file is available. Thus, we do have to download the full data
# set and extract the required stations from it.
# -------------------------------------------------------------------

    # List of required stations (user defines station list)
    #  6660: Zuerich
    # 11120: Innsbruck Airport
    # 11035: Vienna, Hohe Warte
    stations <- c(6660, 11120, 11035)

    # Manually create the output folder if not yet existing.
    dir.create("DWDMOS_S", showWarnings = FALSE)

    # Processing S-type forecasts
    cat("\n\nProcessing S-type DWD MOSMIX forecasts\n")

    # Find latest MOS forecast on opendata.dwd.de
    # As this takes a while: only take the last two entries!
    files <- get_files_available("S", outdir = "DWDMOS_S")
    files <- tail(files, 2)

    # Looping over all not-yet-processed files (if there are any),
    # download the XML file, extract information, done.
    for ( i in 1:nrow(files) ) {

        cat(sprintf("* Downloading \"%s\"\n", as.character(files$srcfile[i])))

        # Downloading the data set
        kmz <- tempfile(pattern = "DWDMOS_S_", fileext = ".kmz")
        check   <- try(download.file(as.character(files$url[i]), kmz, quiet = TRUE))
        if ( inherits(check, "try-error") )
            warning(sprintf("Problems downloading data for %d, skip.", stn))

        # Unzip and parse the XML file
        kml      <- unzip(kmz)
        doc      <- xmlParse(kml)

        # Extracting meta information and datetime information
        meta     <- get_meta_info(doc)
        datetime <- get_datetime(doc)

        # Looping over the stations, extract forecasts and save into
        # the output directory as specified by "dir".
        for ( stn in stations ) {
            fcst <- get_forecasts(sprintf("%05d", stn), doc, datetime, meta)
            write_ascii(fcst, dir = "DWDMOS_S")
        }
        free(doc) # Release connection and free memory

        # Remove the kmz and kml file ...
        file.remove(kmz, kml)
    }











R/plot_dwdforecast.R


#' Default Plot for DWD MOSMIX Forecasts
#'
#' A default plot routine to check the data.
#'
#' @param x object of class \code{dwdforecasts} as returned by
#'     \code{\link{get_forecasts}}.
#'
#' @export
#' @import graphics
# TODO: Write manual page.
plot.dwdforecast <- function(x, ...) {

    # Drop columns containing missing data only
    idx <- which(apply(x, 2, function(x) sum(!is.na(x)) == 0))
    if ( length(idx) > 0 ) x <- x[,-idx]

    # Make time series strictly regular
    if ( ! is.regular(x, strict = TRUE) )
        x <- merge(x, zoo(,seq(min(index(x)), max(index(x)), deltat(x))))

    # Check availability of the different parameters/columns required
    # for the different panels of the default plotting methods.
    # Panels having no data will simply be ignored.
    check_cols <- function(x, params)
        list(N = sum(params %in% names(x)), params = params)
    check <- list(
            temperature = check_cols(x, c("TTT", "Td", "TN", "TX")),
            sunshine    = check_cols(x, c("SunD1", "N", "Neff")),
            prain       = check_cols(x, c("PPPP", "RR1c")),
            wind        = check_cols(x, c("DD", "FF", "FX1"))
        )

    # Take those where at least one parameter is available
    take <- sapply(check, function(x) x$N > 0)
    if ( sum(take) == 0 ) stop("Cannot create a plot, none of the expected parameters available!")

    # Subplot height
    heights <- list(temperature = 2, wind = 2, prain = 1, sunshine = 1)
    heights <- heights[names(take)[take]]

    par(mar = rep(0.1, 4), oma = c(4.1, 7.1, 2.5, 7.2))
    layout(matrix(1L:sum(take), ncol = 1), heights = heights)

    # Function for the baseplot
    baseplot <- function(xlim, ylim, h = NULL, bg = TRUE, ...) {
        plot(NA, xlim = xlim, ylim = ylim, xaxt = "n", xaxs = "i", bty = "n", ...)
        # xblocks
        if ( bg ) {
            b <- seq(as.POSIXct(as.Date(min(xlim))), as.POSIXct(as.Date(max(xlim))+1), by = 86400)
            for ( i in seq(1, length(b)-1, 2) ) {
                zoo::xblocks(c(b[i], b[i+1]), colors$bg)
            }
            # Vertical lines
            v <- seq(as.POSIXct(as.Date(min(xlim))), as.POSIXct(as.Date(max(xlim))+1), by = 12 * 3600)
            abline(v = v, lty = 2, #ifelse(as.POSIXlt(v)$hour == 0, 2, 4),
                          lwd = ifelse(as.POSIXlt(v)$hour == 0, .5, 0.2))
            # Horizontal lines
            if ( is.null(h) ) h <- pretty(ylim)
            abline(h = h, col = colors$gray, lty = 2)
        }
    }

    # X-Limits
    xlim <- range(index(x))

    # A set of colors I'll use
    colors <- list(red = "#FF0000", green = "green", gray = "gray60",
                   blue = "#0000FF",
                   bg = "gray95", orange = "orange")

    # -----------------------------
    # Temperature plot
    if ( check$temperature$N > 0 ) {
        ylim <- range(x[,check$temperature$params] - 273.15, na.rm = TRUE)
        baseplot(xlim, ylim, h = seq(-50, 50, by = 2.5))
        if ( all(c("TTT", "Td") %in% names(x)) )
            add_polygon(x$TTT - 273.15, lower.limit = x$Td - 273.15, col = "#003300")
        print('xxxxxx')
        if ( "E_Td" %in% names(x) )
            add_polygon(              x$Td - 273.15 + x$E_Td/2,
                        lower.limit = x$Td - 273.15 - x$E_Td/2,
                        col = "#00FF00")
        if ( "Td" %in% names(x) )
            lines(x$Td  - 273.15, col = colors$green, lwd = 2)
        if ( "E_TTT" %in% names(x) )
            add_polygon(              x$TTT - 273.15 + x$E_TTT/2,
                        lower.limit = x$TTT - 273.15 - x$E_TTT/2,
                        col = "#FF0000")
        if ( "TTT" %in% names(x) )
            lines(x$TTT - 273.15, col = colors$red,, lwd = 2)
        if ( "TX" %in% names(x) )
            add_steps(x$TX - 273.15, width = 43200, col = colors$red)
        if ( "TN" %in% names(x) )
            add_steps(x$TN - 273.15, width = 43200, col = colors$blue)
        # Text and outer box
        # Text and outer box
        mtext(side = 2, line = 4, "temperatures\n[degrees C]")
        box()
    }

    # -----------------------------
    # Sunshine duration
    if ( check$sunshine$N > 0 ) {
        baseplot(xlim, c(0,60), h = seq(10, 50, by = 10), yaxs = "i")
        if ( "N" %in% names(x) )
            add_polygon(x$N * .6, col = "#303030")
        if ( "Neff" %in% names(x) )
            add_polygon(x$Neff * .6, col = "#303030")
        if ( "SunD1" %in% names(x) )
            add_bars(x$SunD1 / 60, col = colors$orange)
        if ( "SunD" %in% names(x) ) {
            tmp <- na.omit(x$SunD)
            text(as.numeric(index(tmp)) - 43200, 60, col = 1,
                 sprintf("%.1fh/d", tmp / 1440), pos = 1)
        }
        if ( any(c("N", "Neff") %in% names(x)) ) {
            mtext(side = 4, line = 4, "total/effective cloud\ncover [percent]")
            axis(side = 4, at = seq(20, 80, by = 20) * .6,
                 labels = seq(20, 80, by = 20))
        }
        # Text and outer box
        mtext(side = 2, line = 4, "sunshine duration\n[minutes/hour]")
        box()
    }

    # -----------------------------
    # Sunshine duration
    if ( check$prain$N > 0 ) {
        if ( "RR1c" %in% names(x) ) {
            baseplot(xlim, ylim = c(0, max(max(x$RR1c, na.rm = TRUE) * 1.05, 2.5)), yaxs = "i")
            add_bars(x$RR1c, col = colors$blue)
            mtext(side = 4, line = 4, "surface pressure\n[hetcopascal]")
        }
        if ( "PPPP" %in% names(x) ) {
            if ( "RR1c" %in% names(x) ) par(new = TRUE)
            baseplot(xlim, ylim = range(x$PPPP, na.rm = TRUE) / 100, yaxt = "n",
                     bg = FALSE)
            axis(side = 4)
            lines(x$PPPP / 100, lwd = 2)
            mtext(side = 2, line = 4, "precipitation sum\n[km m^-2 / h]")
        }
        # Adding the box
        box()
    }


    # -----------------------------
    # Sunshine duration
    if ( check$wind$N > 0 ) {
        if ( any(c("FF", "FX1") %in% names(x)) ) {
            ylim <- c(0, max(max(range(x[,grep("^(FF|FX1)$", names(x))], na.rm = TRUE) * 1.05), 5))
            baseplot(xlim, ylim, yaxs = "i")
            if ( "FX1" %in% names(x) )
                lines(x$FX1, col = colors$red, lwd = 2)
            if ( "FF" %in% names(x) )
                add_polygon(x$FF, col = colors$blue, lwd = 2)
            box()
            # Adding legend
            if ( all(c("FF", "FX1") %in% names(x)) ) {
                ylab <- "wind speed and maximum\nhourly wind gust [m/s]"
            } else if ( "FF" %in% names(x) ) {
                ylab <- "wind speed\n[m/s]"
            } else {
                ylab <- "maximum hourly\nwind gust [m/s]"
            }
            mtext(side = 2, line = 4, ylab)
        }

        if ( "DD" %in% names(x) ) {
            if ( any(c("FF", "FX1") %in% names(x)) ) par(new = TRUE)
            baseplot(xlim, ylim = c(0, 360), yaxt = "n",
                     bg = ! any(c("FF", "FX1") %in% names(x)))
            points(x$DD, pch = 19, col = rgb(0,0,0,0.8))
            axis(side = 4, at = seq(90, 270, by = 90))

            if ( "E_DD" %in% names(x) ) {
                at <- as.numeric(index(x))
                y0 <- as.numeric(x$DD) - 0.5 * as.numeric(x$E_DD)
                y1 <- as.numeric(x$DD) + 0.5 * as.numeric(x$E_DD)
                segments(at, y0, at, y1,           , col = colors$gray)
                segments(at, y0 + 360, at, y1 + 360, col = colors$gray)
                segments(at, y0 - 360, at, y1 - 360, col = colors$gray)
            }

            mtext(side = 4, line = 4, "wind direction\n[degrees]")
        }

        box()
    }

    # Adding axis
    idx <- as.POSIXct(seq(as.Date(min(xlim)), as.Date(max(xlim)) + 1, by = 1))
    axis(side = 1, at = idx, labels = NA)
    axis(side = 1, at = idx + 43200, labels = strftime(idx, "%b %d"), lwd = 0)

    # Adding title
    if ( "main" %in% names(list(...)) ) { main <- list(...)$main } else {
        main <- sprintf("DWD MOSMIX Forecast, %s (%s)", attr(x, "description"), attr(x, "name"))
    }
    print(main)
    mtext(side = 3, line = 0.5, font = 2, cex = 1.2, main, outer = TRUE)


}

add_steps <- function(x, width, ...) { 
    # Calculate coordinates for the rectangles
    tmp <- na.omit(x)
    segments(as.numeric(index(x)) - width, as.numeric(x),
             as.numeric(index(x)),         as.numeric(x), ...)
}

add_bars <- function(x, col = "yellow", lower.limit = 0) { 
    # Calculate coordinates for the rectangles
    xup <- as.numeric(index(x))
    xlo <- xup - deltat(x)
    ylo <- rep(lower.limit, length(x))
    yup <- as.numeric(x)
    rect(xlo, ylo, xup, yup, col = col, border = NA)
}



#' Add Polygon to Plot
#'
#' Helper function to plot a filled polygon based on a \code{zoo}
#' time series object with nice missing data handling.
#'
#' @param x an univariate \code{zoo} time series object.
#' @param col character, a hex color. Default is \code{"#ff0000"} (red).
#' @param lower.limit numeric, the lower limit used to plot the polygon.
#'        default is \code{0}.
#' @param lwd line width argument.
#'
#' @examples
#' library("zoo")
#' # Create a time series object
#' set.seed(3)
#' a <- zoo(sin(1:100/80*pi) + 3 + rnorm(100, 0, 0.3), 201:300)
#' 
#' # Plot
#' par(mfrow = c(1,3))
#' plot(a, type = "n", main = "Demo Plot 1",
#'      ylim = c(-1, max(a)+1), xaxs = "i", yaxs = "i")
#' add_polygon(a)
#' 
#' # Specify lower.limit to -1 (lower limit of our ylim),
#' # add different color, change line style.
#' plot(a, type = "n", main = "Demo Plot 2",
#'      ylim = c(-1, max(a)+.2), xaxs = "i", yaxs = "i")
#' add_polygon(a, col = "#00CCBB", lower.limit = -1, lwd = 3)
#' 
#' # Using an "upper limit".
#' plot(a, type = "n", main = "Demo Plot 3",
#'      ylim = c(-1, max(a)+.2), xaxs = "i", yaxs = "i")
#' add_polygon(a, col = "#00BBFF", lower.limit = par()$usr[4L])
#' 
#' # Make a copy and add some missing values
#' b <- a
#' b[2:10]  <- NA
#' b[50:55] <- NA
#' b[70]    <- NA
#' 
#' # Plot
#' par(mfrow = c(1,1))
#' 
#' # Same as "Demo Plot 2" with the time series which
#' # contains missing values (b).
#' plot(b, type = "n", main = "Demo Plot 2 With Missing Values",
#'      ylim = c(-1, max(b, na.rm = TRUE)+.2), xaxs = "i", yaxs = "i")
#' add_polygon(b, col = "#00CCBB", lower.limit = -1, lwd = 3)
#' 
#' @import zoo
#' @import graphics
#' @author Reto Stauffer
#' @export
add_polygon <- function(x, col = "#ff0000", lower.limit = 0, lwd = 1) {
    # Need hex color
    if ( ! grepl("^#[A-Za-z0-9]{6}$",col) ) stop("Sorry, need hex color definition for polygon plots.")
    # All elements NA?
    if ( all( is.na(x) ) ) return(invisible(NULL))
    # Else find valid blocks and plot them. Start with 1
    i <- 1
    if ( ! length(lower.limit) == length(x) ) lower.limit <- rep(lower.limit, length(x))

    while ( i <= length(x) ) {

        if ( all(is.na(x)) ) break
        i1 <- min( which( !is.na(x) ) )
        if ( i1 > 1 ) { x <- x[-c(seq(1,i1-1))]; i1 <- 1 }
        # Else check first NA
        if ( ! any(is.na(x)) ) { i2 <- length(x) } else { i2 <- min( which( is.na(x) ) ) - 1 }
        # Create x/y coordinate vectors for the polygon function
#        p_x <- as.numeric(zoo::index(x[i1:i2])); p_x <- c(p_x,max(p_x),min(p_x))
#        p_y <- c(as.numeric(x[i1:i2]),lower.limit, lower.limit )
        p_x <- as.numeric(zoo::index(x[i1:i2])); p_x <- c(p_x,rev(p_x))
        p_y <- c(as.numeric(x[i1:i2]), rev(lower.limit[i1:i2]))
        # Plotting
        graphics::polygon( p_x, p_y, col = sprintf("%s20",col), border = NA )
        graphics::lines( x[i1:i2],   col = col, lwd = lwd )
        # Remove plotted data from time series and continue
        x <- x[-c(i1:i2)]

    }

}

















inst/meteodata_auto.R
# -------------------------------------------------------------------
# - NAME:        automation.R
# - AUTHOR:      Reto Stauffer
# - DATE:        2019-01-09
# -------------------------------------------------------------------
# - DESCRIPTION: Implementation on meteo-data.
# -------------------------------------------------------------------
# - EDITORIAL:   2019-01-09, RS: Created file on thinkreto.
# -------------------------------------------------------------------
# - L@ST MODIFIED: 2019-01-09 21:11 on marvin
# -------------------------------------------------------------------

    library("mosmix")

    # Hotfix for R 3.1
    dir.exists <- function(x, ...) file.exists(x, ...)

    #    coordinates  name                 desc alt
    #  (8.53, 47.48) 06670              ZUERICH 436
    #  (16.37, 48.2) 11034            WIEN/CITY 171
    # (16.37, 48.25) 11035      WIEN/HOHE WARTE 195
    # (16.57, 48.12) 11036             WIEN FL. 183
    #  (13.3, 52.47) 10381        BERLIN-DAHLEM  58
    # (16.42, 48.12) 11040        WIEN/UNTERLAA 201
    # (13.32, 52.57) 10382         BERLIN-TEGEL  37
    #  (13.5, 52.63)  G002          BERLIN-BUCH  60
    #  (13.4, 52.47) 10384     BERLIN-TEMPELHOF  50
    # (11.35, 47.27) 11120        INNSBRUCK FL. 581
    # (13.57, 52.55)  G005       BERLIN-MARZAHN  60
    # (13.52, 52.38) 10385   BERLIN-SCHOENEFELD  47
    # (12.23, 51.42) 10469   LEIPZIG/SCHKEUDITZ 141
    #  (13.73, 52.4)  G006     BERLIN-KANISWALL  33
    # (13.42, 52.52) 10389         BERLIN-ALEX.  37
    # (12.42, 51.32) 10471              LEIPZIG 141
    #  (8.57, 47.38) 06660 ZUERICH (TOWN/VILLE) 556
    # (16.22, 47.83) 11182      WIENER NEUSTADT 285
    #  (9.17, 47.77) K2787 UEBERLINGEN/BODENSEE 440
    stations <- c(6660, 6670, 11034, 11035, 11036, 10381,
                  11040, 10382, 10384, 11120, 10385, 10469,
                  10389, 10471, 11182)


# -------------------------------------------------------------------
# L-type forecasts: XML files are available for each station
# individually. Thus, we only have to download a few kilobytes of
# data. L-type forecasts become available four times a day.
# -------------------------------------------------------------------

    # Processing L-type forecasts (station-by-station)
    cat("\n-------------------------\nProcessing L-type DWD MOSMIX forecasts\n")

    # Looping over the user-specified set of stations
    for ( stn in stations ) {

        # Check for available files on the DWD server
        files <- get_files_available("L", stn, outdir = "DWDMOS_L")

        # No files we have not yet processed? Continue.
        if ( nrow(files) == 0 ) next
    
        # Else loop over all available files and process the data.
        for ( i in 1:nrow(files) ) {

            cat(sprintf("* Downloading \"%s\"\n", as.character(files$srcfile[i])))

            # Downloading the data set
            kmz <- tempfile(pattern = "DWDMOS_L_", fileext = ".kmz")
            check   <- try(download.file(as.character(files$url[i]), kmz,
                                         method = "curl", quiet = TRUE))
            if ( inherits(check, "try-error") )
                warning(sprintf("Problems downloading data for %d, skip.", stn))

            # Unzip and parse the XML file
            kml      <- unzip(kmz)
            doc      <- xmlParse(kml)

            # Extracting meta information and datetime information
            meta     <- get_meta_info(doc)
            datetime <- get_datetime(doc)

            # Extract forecasts for this station and write to ASCII file.
            fcst <- get_forecasts(sprintf("%05d", stn), doc, datetime, meta)
            write_ascii(fcst, dir = "DWDMOS_L")

            XML::free(doc) # Release object, clean up memory

            # Remove the downloaded kmz and extracted kml file.
            file.remove(kmz, kml)
        }
    }
        


# -------------------------------------------------------------------
# S-type forecasts become available 24 times a day. Only one "global"
# XML file is available. Thus, we do have to download the full data
# set and extract the required stations from it.
# -------------------------------------------------------------------

    # Processing S-type forecasts
    cat("\n-------------------------\nProcessing S-type DWD MOSMIX forecasts\n")

    # Find latest MOS forecast on opendata.dwd.de
    # As this takes a while: only take the last two entries!
    files <- get_files_available("S", outdir = "DWDMOS_S")

    # Looping over all not-yet-processed files (if there are any),
    # download the XML file, extract information, done.
    if ( nrow(files) > 0 ) {
        cat(sprintf("* Found %d files to process ...\n", nrow(files)))
        for ( i in 1:nrow(files) ) {
   
            cat(sprintf("* Downloading \"%s\"\n", as.character(files$srcfile[i])))
   
            # Downloading the data set
            kmz <- tempfile(pattern = "DWDMOS_S_", fileext = ".kmz")
            check   <- try(download.file(as.character(files$url[i]), kmz,
                                         method = "curl", quiet = TRUE))
            if ( inherits(check, "try-error") )
                warning(sprintf("Problems downloading data for %d, skip.", stn))
   
            # Unzip and parse the XML file
            kml      <- unzip(kmz)
            doc      <- xmlParse(kml)
   
            # Extracting meta information and datetime information
            meta     <- get_meta_info(doc)
            datetime <- get_datetime(doc)
   
            # Looping over the stations, extract forecasts and save into
            # the output directory as specified by "dir".
            for ( stn in stations ) {
                fcst <- get_forecasts(sprintf("%05d", stn), doc, datetime, meta)
                write_ascii(fcst, dir = "DWDMOS_S")
            }
            XML::free(doc) # Release object, clean up memory
   
            # Remove the kmz and kml file ...
            file.remove(kmz, kml)
        }
    } else {
        cat(sprintf("* No new files, nothing to do at the moment.\n"))
    }







